### PR TITLE
Update data modeling developer guide

### PR DESCRIPTION
**Pull request by yufeih**
Added section about how to model data in docfx

### CODE REVIEW

File: docs/engineering/developer-guide.md [hunk 1]
```diff
@@ -67,6 +67,41 @@ Write complex functions by composing smaller, simpler functions. Write functions
...
+
+When YAML is used, we only support a subset of YAML that is JSON compatible, features like multiple documents, non-scalar keys, anchors and references are not supported.
+
+Flattened structure is prefered over nested structure, because YAML is indention based, nested structure creates a very bad YAML authoring experience.
+
+The default JSON naming convention is *snake_case* for property names and enum values:
+
+```json
+{
+    "a_property_name": "a_property_value"
```
Comment (OsmondJiang) [inline docs/engineering/developer-guide.md]: enum property value

Comment (OsmondJiang) [reply docs/engineering/developer-guide.md]: applies to input and output models

File: docs/engineering/developer-guide.md [hunk 2]
```diff
@@ -67,6 +67,41 @@ Write complex functions by composing smaller, simpler functions. Write functions
...
+}
+```
+
+`null`s are ignored using `[NullValueHandling.Ignore]` to void [the billion-dollar mistake](https://en.wikipedia.org/wiki/Tony_Hoare). We will be using [strict null checking](https://blogs.msdn.microsoft.com/dotnet/2017/11/15/nullable-reference-types-in-csharp/) when C# 8 arrives.
```
Comment (OsmondJiang) [inline docs/engineering/developer-guide.md]: extensions

File: docs/engineering/developer-guide.md [hunk 3]
```diff
@@ -67,6 +67,41 @@ Write complex functions by composing smaller, simpler functions. Write functions
...
+    - With no methods that contains business logic
+
+- Avoid `[JsonProperty]` unless the property does not use the above naming convention for backward compatibility reasons:
+    - It ensures we provide consistent input and output user experience.
+    - It ensures our C# model directly mapps to data contract using the same name.
+    - It is easier to find all naming exceptions by searching `[JsonProperty]`.
+
+- Avoid `[JsonIgnore]`: use of `[JsonIgnore]` typically means that you are mixing logic with data. You can use tuples to pass `JsonIgnored` parameters.
+
+- Avoid inheritance: type information does not exists in wire format, use enums instead.
```
Comment (OsmondJiang) [inline docs/engineering/developer-guide.md]: 多态

File: docs/developer-guide.md [hunk 4]
```diff
@@ -79,6 +79,43 @@ Write complex functions by composing smaller, simpler functions. Write functions
...
+
+> Note that `null`s for unknown properties marked as `[JsonExtensionData]` are still preserved.
+
+- Make data contract [POCO (plain Old C# Object)](https://stackoverflow.com/questions/250001/poco-definition):
+    - With only simple serializable properties
+    - With no methods that contains business logic
+
+- Avoid `[JsonProperty]` unless the property does not use the above naming convention for backward compatibility reasons:
+    - It ensures we provide consistent input and output user experience.
+    - It ensures our C# model directly mapps to data contract using the same name.
```
Comment (live1206) [inline docs/developer-guide.md]:
```suggestion
    - It ensures our C# model directly maps to data contract using the same name.
```

File: docs/developer-guide.md [hunk 5]
```diff
@@ -79,6 +79,43 @@ Write complex functions by composing smaller, simpler functions. Write functions
...
 
+### Data Modeling
+
+JSON `(.json)` is the format of our data contract. YAML `(.yml)` is an alternative and preferred input format for config and authoring.
+
+Whenever JSON is supported as an input format, YAML is supported as well. The same is true vice-versa.
+
+When YAML is used, we only support a subset of YAML that is JSON compatible, features like multiple documents, non-scalar keys, anchors and references are not supported.
+
+The following guidelines applies to both input models and output models:
```
Comment (live1206) [inline docs/developer-guide.md]:
```suggestion
The following guidelines apply to both input models and output models:
```

**Pull request merged**